File: utils/vfs/initrd/create_initrd.py
# **************************************************************************** #
#                                                                              #
#                                                         :::      ::::::::    #
#    create_initrd.py                                   :+:      :+:    :+:    #
#                                                     +:+ +:+         +:+      #
#    By: vvaucoul <vvaucoul@student.42.fr>          +#+  +:+       +#+         #
#                                                 +#+#+#+#+#+   +#+            #
#    Created: 2024/01/13 13:56:03 by vvaucoul          #+#    #+#              #
#    Updated: 2024/01/13 14:19:21 by vvaucoul         ###   ########.fr        #
#                                                                              #
# **************************************************************************** #

import struct
import os

# Definition of the initrd header structure
InitrdHeader = struct.Struct('I')  # uint32_t nfiles
InitrdFileHeader = struct.Struct('B64sII')  # uint8_t magic; char name[64]; uint32_t size; uint32_t offset

# Directory containing the files to include in the initrd
initrd_directory = 'initrd_files'

# Initrd magic
magic = 0xBF

# Create a list to store the file headers
file_headers = []
files_data = b''

# Iterate through the files in the initrd directory
for filename in os.listdir(initrd_directory):
    filepath = os.path.join(initrd_directory, filename)
    if os.path.isfile(filepath):
        with open(filepath, 'rb') as f:
            file_data = f.read()
            files_data += file_data
            # Add a magic byte in front of the file name
            file_headers.append((magic, filename.encode('ascii'), len(file_data), len(files_data) - len(file_data)))

# Create the initrd header
initrd_header = InitrdHeader.pack(len(file_headers))

# Create the initrd file headers
headers_data = b''.join(InitrdFileHeader.pack(*header) for header in file_headers)

# Write the complete initrd to a file
with open('initrd.img', 'wb') as f:
    f.write(initrd_header)
    f.write(headers_data)
    f.write(files_data)

File: utils/vfs/initrd/create_initrd.c
/* ************************************************************************** */
/*                                                                            */
/*                                                        :::      ::::::::   */
/*   create_initrd.c                                    :+:      :+:    :+:   */
/*                                                    +:+ +:+         +:+     */
/*   By: vvaucoul <vvaucoul@student.42.fr>          +#+  +:+       +#+        */
/*                                                +#+#+#+#+#+   +#+           */
/*   Created: 2024/01/16 13:41:12 by vvaucoul          #+#    #+#             */
/*   Updated: 2024/01/16 16:04:16 by vvaucoul         ###   ########.fr       */
/*                                                                            */
/* ************************************************************************** */

#include <dirent.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/stat.h>

#define INITRD_FILE_SIZE 64

typedef struct {
    uint32_t nfiles;
} __attribute__((packed)) InitrdHeader;

typedef struct {
    uint8_t magic;
    char name[INITRD_FILE_SIZE];
    uint32_t size;
    uint32_t offset;
} __attribute__((packed)) InitrdFileHeader;

int main() {
    DIR *dir;
    struct dirent *ent;
    if ((dir = opendir("initrd_files")) != NULL) {
        int file_count = 0;
        while ((ent = readdir(dir)) != NULL) {
            if (ent->d_type == DT_REG) { // If the entry is a regular file
                file_count++;
            }
        }
        closedir(dir);

        InitrdHeader header;
        header.nfiles = file_count;

        FILE *initrd = fopen("initrd.img", "wb");
        fwrite(&header, sizeof(InitrdHeader), 1, initrd);

        uint32_t offset = sizeof(InitrdHeader) + sizeof(InitrdFileHeader) * header.nfiles;

        // First pass: write headers
        dir = opendir("initrd_files");
        while ((ent = readdir(dir)) != NULL) {
            if (ent->d_type == DT_REG) {
                InitrdFileHeader file_header;
                file_header.magic = 0xBF;
                strncpy(file_header.name, ent->d_name, INITRD_FILE_SIZE);

                struct stat st;
                char filepath[512];
                snprintf(filepath, sizeof(filepath), "initrd_files/%s", ent->d_name);
                stat(filepath, &st);

                file_header.size = st.st_size;
                file_header.offset = offset;

                fwrite(&file_header, sizeof(InitrdFileHeader), 1, initrd);

                offset += file_header.size;
            }
        }
        closedir(dir);

        // Second pass: write file contents
        dir = opendir("initrd_files");
        while ((ent = readdir(dir)) != NULL) {
            if (ent->d_type == DT_REG) {
                char filepath[512];
                snprintf(filepath, sizeof(filepath), "initrd_files/%s", ent->d_name);
                FILE *file = fopen(filepath, "rb");

                struct stat st;
                stat(filepath, &st);
                char *buffer = malloc(st.st_size);
                fread(buffer, st.st_size, 1, file);
                fwrite(buffer, st.st_size, 1, initrd);

                free(buffer);
                fclose(file);
            }
        }
        closedir(dir);

        fclose(initrd);

    } else {
        perror("Could not open directory initrd_files");
        return EXIT_FAILURE;
    }

    return 0;
}